src/dune_engine/fs_memo.ml

```diff
@@ -2,39 +2,93 @@ open! Stdune
+   calls [depend] and uses the corresponding path in the right order. *)
 let depend path =
   if Path.is_in_build_dir path then
     Code_error.raise "Fs_memo.depend called on a build path" [];
   Memo.exec memo path
 
-let file_exists path =
-  let+ () = depend path in
-  Path.exists path
+(* Declare a dependency on [path] and then sample the current value using the
+   supplied function [f].
 
-let file_digest path =
+   If the order is reversed, the value can change after we sample it but before
+   we register the dependency, which can result in memoizing a stale value. This
+   scenario is purely hypothetical (at least for now) but it's nice to rule it
+   out explicitly by doing things in the right order.
+
+   Currently, we do not expose this low-level primitive. If you need it, perhaps
+   you could add a higher-level primitive instead, such as [file_exists]? *)
+let declaring_dependency path ~f =
   let+ () = depend path in
-  Cached_digest.source_or_external_file path
+  f path
+
+let file_exists = declaring_dependency ~f:Path.exists
+
+(* CR-someday amokhov: It is unclear if we got the layers of abstraction right
+   here. One could argue that caching is a higher-level concept compared to file
+   watching, and we should expose this function from the [Cached_digest] module
+   instead. For now, we keep it here because it seems nice to group all tracked
+   file system access functions in one place, and exposing an uncached version
+   of [file_digest] seems error-prone. We may need to rethink this decision. *)
+let file_digest = declaring_dependency ~f:Cached_digest.source_or_external_file
+
+let dir_contents = declaring_dependency ~f:Path.readdir_unsorted_with_kinds
+
+module Event = struct
+  type kind =
+    | File_created
+    | File_deleted
+    | File_changed
+    | Directory_created
+    | Directory_deleted
 
-module Invalidate_result = struct
   type t =
-    | Invalidated
-    | Skipped
+    { path : Path.t
+    ; kind : kind
+    }
+
+  let create ~kind ~path =
+    if Path.is_in_build_dir path then
+      Code_error.raise "Fs_memo.Event.create called on a build path" [];
+    { path; kind }
 end
 
-let invalidate path =
-  if Path.is_in_build_dir path then
-    Code_error.raise "Fs_memo.invalidate called on a build path" [];
-  match Memo.Expert.previously_evaluated_cell memo path with
-  | None -> Invalidate_result.Skipped
-  | Some cell ->
-    Memo.Cell.invalidate cell;
-    Invalidated
+module Rebuild_required = struct
+  type t =
+    | Yes
+    | No
+end
+
+let process_events events =
+  (* Knowing that the list is non-empty makes it easier to think about the logic
+     that checks the [Memo.incremental_mode_enabled] flag. *)
+  let events = Nonempty_list.to_list events in
+  let rebuild_required =
+    (* We can't use [List.exists] here due to its short-circuiting behaviour. *)
+    List.fold_left events ~init:false ~f:(fun acc { Event.path; _ } ->
```

Comment:
It feels like the accumulator could be a `Rebuild_required.t` directly here since the boolean is immediately converted to one just after.

Reply:
Indeed. Fixing!